File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/d466288ea1696c8492238b770d50afca.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

typedef struct {
    int* arr;
    int sign;
} Int;

int size;
int max = 1000000000;

Int* createInt(const char* a);
Int* copyInt(const Int* a);
void destroyInt(Int* a);

Int* assignInt(Int* a, const Int* b);
Int* assignString(Int* a, const char* b);

int compareInt(const Int* a, const Int* b);
int compareString(const Int* a, const char* b);

Int* negateInt(const Int* a, Int* result);

Int* addInt(const Int* a, const Int* b, Int* result);
Int* addString(const Int* a, const char* b, Int* result);

Int* subtractInt(const Int* a, const Int* b, Int* result);
Int* subtractString(const Int* a, const char* b, Int* result);

Int* multiplyInt(const Int* a, const Int* b, Int* result);
Int* multiplyString(const Int* a, const char* b, Int* result);

char* toString(const Int* a);

void set_size(int k);
int num(const char* a);
char* str(int a);
Int* absInt(const Int* a);

int qanak;

Int* solve(const char* s, int h);
void mainFunction();

int main() {
    qanak = 0;
    set_size(12);
    Int* answer = createInt("0");
    char s[100];
    scanf("%s", s);
    if (strlen(s) == 1) {
        if (s[0] == '?' || (s[0] >= 'A' && s[0] <= 'Z')) {
            assignString(answer, "9");
        } else {
            assignString(answer, "1");
        }
        printf("%s\n", toString(answer));
        destroyInt(answer);
        return 0;
    }
    if (s[0] == '?') {
        char st[100];
        strcpy(st, s);
        memmove(st, st + 1, strlen(st));
        Int* temp = solve(st, 0);
        multiplyString(temp, "9", answer);
        destroyInt(temp);
    } else if (s[0] >= 'A' && s[0] <= 'Z') {
        for (int k = 1; k < 10; ++k) {
            char st[100];
            strcpy(st, s);
            qanak = 0;
            for (int i = 0; i < strlen(st); ++i) {
                if (st[i] == s[0]) {
                    st[i] = (char)('0' + k);
                }
            }
            Int* temp = solve(st, 1);
            addInt(answer, temp, answer);
            destroyInt(temp);
        }
    } else {
        solve(s, 0);
        multiplyString(answer, "9", answer);
    }
    printf("%s", toString(answer));
    for (int i = 0; i < qanak; ++i) {
        printf("0");
    }
    printf("\n");
    destroyInt(answer);
    return 0;
}

Int* createInt(const char* a) {
    Int* num = (Int*)malloc(sizeof(Int));
    num->arr = (int*)calloc(size + 1, sizeof(int));
    num->sign = 0;
    if (a[0] == '-') {
        num->sign = 1;
        memmove(a, a + 1, strlen(a));
    }
    num->arr[0] = 0;
    while (strlen(a) > 9) {
        num->arr[0]++;
        num->arr[num->arr[0]] = num(a + strlen(a) - 9);
        memmove(a + strlen(a) - 9, a, strlen(a));
        a[strlen(a) - 9] = '\0';
    }
    num->arr[0]++;
    num->arr[num->arr[0]] = num(a);
    return num;
}

Int* copyInt(const Int* a) {
    Int* num = (Int*)malloc(sizeof(Int));
    num->arr = (int*)calloc(size + 1, sizeof(int));
    num->sign = a->sign;
    num->arr[0] = a->arr[0];
    for (int i = 1; i <= num->arr[0]; i++) {
        num->arr[i] = a->arr[i];
    }
    return num;
}

void destroyInt(Int* a) {
    free(a->arr);
    free(a);
}

Int* assignInt(Int* a, const Int* b) {
    if (a != b) {
        free(a->arr);
        a->arr = (int*)calloc(size + 1, sizeof(int));
        a->sign = b->sign;
        a->arr[0] = b->arr[0];
        for (int i = 1; i <= a->arr[0]; i++) {
            a->arr[i] = b->arr[i];
        }
    }
    return a;
}

Int* assignString(Int* a, const char* b) {
    Int* num = createInt(b);
    assignInt(a, num);
    destroyInt(num);
    return a;
}

int compareInt(const Int* a, const Int* b) {
    int* brr = b->arr;
    if (a->sign != b->sign) {
        return a->sign < b->sign ? -1 : 1;
    }
    if (a->sign == 0) {
        if (a->arr[0] != brr[0]) {
            return a->arr[0] > brr[0] ? 1 : -1;
        }
        for (int i = a->arr[0]; i >= 1; i--) {
            if (a->arr[i] != brr[i]) {
                return a->arr[i] > brr[i] ? 1 : -1;
            }
        }
        return 0;
    }
    if (a->arr[0] != brr[0]) {
        return a->arr[0] < brr[0] ? 1 : -1;
    }
    for (int i = a->arr[0]; i >= 1; i--) {
        if (a->arr[i] != brr[i]) {
            return a->arr[i] < brr[i] ? 1 : -1;
        }
    }
    return 0;
}

int compareString(const Int* a, const char* b) {
    Int* num = createInt(b);
    int result = compareInt(a, num);
    destroyInt(num);
    return result;
}

Int* negateInt(const Int* a, Int* result) {
    assignInt(result, a);
    result->sign = (result->sign + 1) % 2;
    return result;
}

Int* addInt(const Int* a, const Int* b, Int* result) {
    int* brr = b->arr;
    int carry = 0;
    if (compareString(a, "0") < 0) {
        if (compareString(b, "0") < 0) {
            if (a->arr[0] >= brr[0]) {
                for (int i = 1; i <= brr[0]; i++) {
                    result->arr[i] = a->arr[i] + brr[i] + carry;
                    carry = result->arr[i] / max;
                    result->arr[i] %= max;
                }
                while (carry) {
                    result->arr[i] = result->arr[i] + carry;
                    carry = result->arr[i] / max;
                    result->arr[i] %= max;
                    i++;
                }
                if (i == result->arr[0] + 2) {
                    result->arr[0]++;
                }
            } else {
                Int* c = copyInt(a);
                Int* d = copyInt(b);
                assignInt(result, addInt(d, c, result));
                destroyInt(c);
                destroyInt(d);
            }
        } else {
            if (compareInt(a, negateInt(b, result)) == 0) {
                assignString(result, "0");
            } else if (compareInt(negateInt(a, result), b) > 0) {
                for (int i = 1; i <= brr[0]; i++) {
                    result->arr[i] = a->arr[i] - brr[i] + carry;
                    if (result->arr[i] < 0) {
                        carry = -1;
                        result->arr[i] += max;
                    } else {
                        carry = 0;
                    }
                }
                while (carry) {
                    result->arr[i] = result->arr[i] + carry;
                    if (result->arr[i] < 0) {
                        carry = -1;
                        result->arr[i] += max;
                    } else {
                        carry = 0;
                    }
                    i++;
                }
                while (result->arr[result->arr[0]] == 0 && result->arr[0] > 1) {
                    result->arr[0]--;
                }
            } else {
                Int* c = copyInt(a);
                Int* d = copyInt(b);
                assignInt(result, subtractInt(d, c, result));
                destroyInt(c);
                destroyInt(d);
            }
        }
    } else {
        if (compareString(b, "0") < 0) {
            if (compareInt(a, negateInt(b, result)) == 0) {
                assignString(result, "0");
            } else if (compareInt(a, negateInt(b, result)) > 0) {
                for (int i = 1; i <= brr[0]; i++) {
                    result->arr[i] = a->arr[i] - brr[i] + carry;
                    if (result->arr[i] < 0) {
                        carry = -1;
                        result->arr[i] += max;
                    } else {
                        carry = 0;
                    }
                }
                while (carry) {
                    result->arr[i] = result->arr[i] + carry;
                    if (result->arr[i] < 0) {
                        carry = -1;
                        result->arr[i] += max;
                    } else {
                        carry = 0;
                    }
                    i++;
                }
                while (result->arr[result->arr[0]] == 0 && result->arr[0] > 1) {
                    result->arr[0]--;
                }
            } else {
                Int* c = copyInt(a);
                Int* d = copyInt(b);
                assignInt(result, subtractInt(d, c, result));
                destroyInt(c);
                destroyInt(d);
            }
        } else {
            Int* c = negateInt(a, result);
            Int* d = negateInt(b, result);
            assignInt(result, addInt(d, c, result));
            destroyInt(c);
            destroyInt(d);
        }
    }
    return result;
}

Int* addString(const Int* a, const char* b, Int* result) {
    Int* num = createInt(b);
    addInt(a, num, result);
    destroyInt(num);
    return result;
}

Int* subtractInt(const Int* a, const Int* b, Int* result) {
    Int* c = negateInt(b, result);
    addInt(a, c, result);
    destroyInt(c);
    return result;
}

Int* subtractString(const Int* a, const char* b, Int* result) {
    Int* num = createInt(b);
    subtractInt(a, num, result);
    destroyInt(num);
    return result;
}

Int* multiplyInt(const Int* a, const Int* b, Int* result) {
    if (compareString(b, "0") == 0) {
        assignString(result, "0");
    } else if (compareString(b, str(max)) == 0) {
        if (compareString(a, "0") != 0) {
            for (int i = a->arr[0]; i >= 1; i--) {
                result->arr[i + 1] = a->arr[i];
            }
            result->arr[1] = 0;
            result->arr[0]++;
        }
    } else if (compareInt(absInt(b), str(max)) < 0) {
        long long tmp;
        int carry = 0;
        int* brr = b->arr;
        for (int i = 1; i <= a->arr[0]; i++) {
            tmp = (long long)a->arr[i] * (long long)brr[1] + (long long)carry;
            result->arr[i] = tmp % (long long)max;
            carry = tmp / (long long)max;
        }
        if (carry) {
            result->arr[result->arr[0] + 1] = carry;
            result->arr[0]++;
        }
    } else {
        Int* tmp = createInt("0");
        Int* res = createInt("0");
        int* brr = b->arr;
        for (int i = brr[0]; i >= 1; i--) {
            multiplyString(a, str(brr[i]), tmp);
            multiplyString(res, str(max), res);
            addInt(res, tmp, res);
        }
        assignInt(result, res);
        destroyInt(tmp);
        destroyInt(res);
    }
    result->sign = (a->sign + b->sign) % 2;
    return result;
}

Int* multiplyString(const Int* a, const char* b, Int* result) {
    Int* num = createInt(b);
    multiplyInt(a, num, result);
    destroyInt(num);
    return result;
}

char* toString(const Int* a) {
    char* res = (char*)malloc(1000 * sizeof(char));
    int j, k;
    int* brr = a->arr;
    if (a->sign) {
        res[0] = '-';
        res[1] = '\0';
    } else {
        sprintf(res, "%d", brr[brr[0]]);
        for (int i = brr[0] - 1; i >= 1; i--) {
            k = max;
            for (j = 0; j < 9; j++) {
                sprintf(res + strlen(res), "%d", (brr[i] % k) / (k /= 10));
            }
        }
    }
    return res;
}

void set_size(int k) {
    size = k;
}

int num(const char* a) {
    int res = 0;
    for (int i = 0; i < strlen(a); i++) {
        res = res * 10 + (a[i] - '0');
    }
    return res;
}

char* str(int a) {
    char* res = (char*)malloc(1000 * sizeof(char));
    int h = (a < 0 ? 1 : 0);
    a = (a < 0 ? -a : a);
    int i = 0;
    while (a != 0) {
        res[i++] = (char)(a % 10 + '0');
        a /= 10;
    }
    if (i == 0) {
        res[i++] = '0';
    }
    if (h) {
        res[i++] = '-';
    }
    res[i] = '\0';
    return res;
}

Int* absInt(const Int* a) {
    Int* result = copyInt(a);
    if (compareString(result, "0") < 0) {
        negateInt(result, result);
    }
    return result;
}

int qanak;

Int* solve(const char* s, int h) {
    bool u[26] = { false };
    Int* answer = createInt("1");
    int k = 0;
    for (int i = 0; i < strlen(s); ++i) {
        if (s[i] == '?') {
            k++;
        }
        if (s[i] >= 'A' && s[i] <= 'Z') {
            u[s[i] - 'A'] = true;
        }
    }
    for (int i = 0; i < k; ++i) {
        qanak++;
    }
    k = 0;
    for (int i = 0; i < 26; ++i) {
        if (u[i]) {
            k++;
        }
    }
    if (k > 10) {
        return createInt("0");
    }
    for (int i = 0; i < k; ++i) {
        multiplyString(answer, str(10 - i - h), answer);
    }
    return answer;
}

void mainFunction() {
    qanak = 0;
    set_size(12);
    Int* answer = createInt("0");
    char s[100];
    scanf("%s", s);
    if (strlen(s) == 1) {
        if (s[0] == '?' || (s[0] >= 'A' && s[0] <= 'Z')) {
            assignString(answer, "9");
        } else {
            assignString(answer, "1");
        }
        printf("%s\n", toString(answer));
        destroyInt(answer);
        return;
    }
    if (s[0] == '?') {
        char st[100];
        strcpy(st, s);
        memmove(st, st + 1, strlen(st));
        Int* temp = solve(st, 0);
        multiplyString(temp, "9", answer);
        destroyInt(temp);
    } else if (s[0] >= 'A' && s[0] <= 'Z') {
        for (int k = 1; k < 10; ++k) {
            char st[100];
            strcpy(st, s);
            qanak = 0;
            for (int i = 0; i < strlen(st); ++i) {
                if (st[i] == s[0]) {
                    st[i] = (char)('0' + k);
                }
            }
            Int* temp = solve(st, 1);
            addInt(answer, temp, answer);
            destroyInt(temp);
        }
    } else {
        solve(s, 0);
        multiplyString(answer, "9", answer);
    }
    printf("%s", toString(answer));
    for (int i = 0; i < qanak; ++i) {
        printf("0");
    }
    printf("\n");
    destroyInt(answer);
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/d466288ea1696c8492238b770d50afca.py
class Int:
    size = 0
    max = 1000000000

    def __init__(self, a="0"):
        self.arr = [0] * (Int.size + 1)
        self.sign = 0
        if a[0] == '-':
            self.sign = 1
            a = a[1:]
        self.arr[0] = 0
        while len(a) > 9:
            self.arr[0] += 1
            self.arr[self.arr[0]] = Int.num(a[-9:])
            a = a[:-9]
        self.arr[0] += 1
        self.arr[self.arr[0]] = Int.num(a)

    def __str__(self):
        res = ""
        if self.sign:
            res += "-"
        res += Int.str(self.arr[self.arr[0]])
        for i in range(self.arr[0] - 1, 0, -1):
            k = Int.max
            for j in range(9):
                res += Int.str((self.arr[i] % k) // (k // 10))
                k //= 10
        return res

    def __mul__(self, a):
        if a == "0":
            return Int("0")
        elif a == str(Int.max):
            if self != "0":
                self.arr.append(0)
                self.arr[0] += 1
        elif Int.abs(a) < str(Int.max):
            tmp = 0
            carry = 0
            brr = a.arr
            for i in range(1, self.arr[0] + 1):
                tmp = self.arr[i] * brr[1] + carry
                self.arr[i] = tmp % Int.max
                carry = tmp // Int.max
            if carry:
                self.arr.append(carry)
                self.arr[0] += 1
        else:
            tmp = Int()
            res = Int("0")
            brr = a.arr
            for i in range(brr[0], 0, -1):
                tmp = self * Int.str(brr[i])
                res = res * Int.str(Int.max) + tmp
            self.arr = res.arr
        self.sign = (self.sign + a.sign) % 2
        return self

    def __imul__(self, a):
        return self * a

    def __rmul__(self, a):
        return self * a

    def __add__(self, a):
        brr = a.arr
        carry = 0
        if self < "0":
            if a < "0":
                if self.arr[0] >= brr[0]:
                    for i in range(1, brr[0] + 1):
                        self.arr[i] = self.arr[i] + brr[i] + carry
                        carry = self.arr[i] // Int.max
                        self.arr[i] %= Int.max
                    while carry:
                        self.arr[i] = self.arr[i] + carry
                        carry = self.arr[i] // Int.max
                        self.arr[i] %= Int.max
                        i += 1
                    if i == self.arr[0] + 2:
                        self.arr[0] += 1
                else:
                    c = Int(self)
                    d = Int(a)
                    self = d + c
            else:
                if self == (-a):
                    self = "0"
                elif -self > a:
                    for i in range(1, brr[0] + 1):
                        self.arr[i] = self.arr[i] - brr[i] + carry
                        if self.arr[i] < 0:
                            carry = -1
                            self.arr[i] += Int.max
                        else:
                            carry = 0
                    while carry:
                        self.arr[i] = self.arr[i] + carry
                        if self.arr[i] < 0:
                            carry = -1
                            self.arr[i] += Int.max
                        else:
                            carry = 0
                        i += 1
                    while self.arr[self.arr[0]] == 0 and self.arr[0] > 1:
                        self.arr[0] -= 1
                else:
                    c = Int(self)
                    d = Int(a)
                    self = d + c
        else:
            if a < "0":
                if self == -a:
                    self = "0"
                elif self > -a:
                    for i in range(1, brr[0] + 1):
                        self.arr[i] = self.arr[i] - brr[i] + carry
                        if self.arr[i] < 0:
                            carry = -1
                            self.arr[i] += Int.max
                        else:
                            carry = 0
                    while carry:
                        self.arr[i] = self.arr[i] + carry
                        if self.arr[i] < 0:
                            carry = -1
                            self.arr[i] += Int.max
                        else:
                            carry = 0
                        i += 1
                    while self.arr[self.arr[0]] == 0 and self.arr[0] > 1:
                        self.arr[0] -= 1
                else:
                    c = Int(self)
                    d = Int(a)
                    self = d + c
            else:
                c = -Int(self)
                d = -Int(a)
                self = -(d + c)
        return self

    def __iadd__(self, a):
        return self + a

    def __radd__(self, a):
        return self + a

    def __sub__(self, a):
        return self + (-a)

    def __isub__(self, a):
        return self - a

    def __rsub__(self, a):
        return -self + a

    def __neg__(self):
        c = Int(self)
        c.sign = (c.sign + 1) % 2
        return c

    def __eq__(self, a):
        return not (a > self or a < self)

    def __ne__(self, a):
        return not (self == a)

    def __lt__(self, a):
        return a > self

    def __le__(self, a):
        return not (self > a)

    def __gt__(self, a):
        brr = a.arr
        if self.sign != a.sign:
            return self.sign < a.sign
        if self.sign == 0:
            if self.arr[0] != brr[0]:
                return self.arr[0] > brr[0]
            for i in range(self.arr[0], 0, -1):
                if self.arr[i] != brr[i]:
                    return self.arr[i] > brr[i]
            return False
        if self.arr[0] != brr[0]:
            return self.arr[0] < brr[0]
        for i in range(self.arr[0], 0, -1):
            if self.arr[i] != brr[i]:
                return self.arr[i] < brr[i]
        return False

    def __ge__(self, a):
        return not (self < a)

    @staticmethod
    def abs(a):
        return -a if a < "0" else a

    @staticmethod
    def num(a):
        res = 0
        for i in range(len(a)):
            res = res * 10 + int(a[i])
        return res

    @staticmethod
    def str(a):
        res = ""
        h = 1 if a < 0 else 0
        a = -a if a < 0 else a
        while a != 0:
            res = str(a % 10) + res
            a //= 10
        if len(res) == 0:
            res = "0"
        if h:
            res = "-" + res
        return res

    @staticmethod
    def set_size(k):
        Int.size = k


def solve(s, h=0):
    u = [False] * 26
    answer = Int("1")
    k = 0
    for i in range(len(s)):
        if s[i] == '?':
            k += 1
        if s[i] >= 'A' and s[i] <= 'Z':
            u[ord(s[i]) - ord('A')] = True
    for i in range(k):
        global qanak
        qanak += 1
    k = 0
    for i in range(26):
        if u[i]:
            k += 1
    if k > 10:
        return Int("0")
    for i in range(k):
        answer *= Int.str(10 - i - h)
    return answer


qanak = 0
Int.set_size(12)
answer = Int("0")
s = input()
if len(s) == 1:
    if s[0] == '?' or (s[0] >= 'A' and s[0] <= 'Z'):
        answer = Int("9")
    else:
        answer = Int("1")
    print(answer)
    exit()
if s[0] == '?':
    st = s[1:]
    answer = Int("9") * solve(st)
elif s[0] >= 'A' and s[0] <= 'Z':
    for k in range(1, 10):
        st = s
        qanak = 0
        for i in range(len(st)):
            if st[i] == s[0]:
                st = st[:i] + str(k) + st[i + 1:]
        answer += solve(st, 1)
else:
    answer = solve(s)
print(answer, end="")
for i in range(qanak):
    print("0", end="")
print()
